consensus/istanbul/core/backlog.go

```diff
@@ -29,189 +35,314 @@ var (
+	c.currentView = view
 
-		logger := c.newLogger("func", "processBacklog", "from", src)
-		isFuture := false
-
-		// We stop processing if
-		//   1. backlog is empty
-		//   2. The first message in queue is a future message
-		for !(backlog.Empty() || isFuture) {
-			m, prio := backlog.Pop()
-			msg := m.(*istanbul.Message)
-			var view *istanbul.View
-			switch msg.Code {
-			case istanbul.MsgPreprepare:
-				var m *istanbul.Preprepare
-				err := msg.Decode(&m)
-				if err == nil {
-					view = m.View
-				}
-			case istanbul.MsgPrepare:
-				var sub *istanbul.Subject
-				err := msg.Decode(&sub)
-				if err == nil {
-					view = sub.View
-				}
-			case istanbul.MsgCommit:
-				var cs *istanbul.CommittedSubject
-				err := msg.Decode(&cs)
-				if err == nil {
-					view = cs.Subject.View
+	c.processBacklog()
+}
+
+func (c *msgBacklogImpl) processBacklog() {
+	for _, seq := range c.getSortedBacklogSeqs() {
+
+		logger := c.logger.New("func", "processBacklog", "current_view", c.currentView, "current_state", c.currentState.String(), "for_seq", seq)
+
+		if seq < c.currentView.Sequence.Uint64() {
+			// Earlier sequence. Prune all messages.
+			c.clearBacklogForSeq(seq)
+		} else if seq == c.currentView.Sequence.Uint64() {
+			// Current sequence. Process all in order.
+			c.processBacklogForSeq(seq, func(msg *istanbul.Message) bool {
+				view, err := extractMessageView(msg)
+				if err != nil {
+					logger.Error("Error decoding msg", "msg", msg, "err", err)
+					return false
 				}
-			case istanbul.MsgRoundChange:
-				var rc *istanbul.RoundChange
-				err := msg.Decode(&rc)
-				if err == nil {
-					view = rc.View
+				if view == nil {
+					logger.Error("Nil view", "msg", msg)
+					return false
 				}
-			}
 
-			if view == nil {
-				logger.Debug("Nil view", "msg", msg)
-				continue
-			}
+				err = c.checkMessage(msg.Code, view)
 
-			// Push back if it's a future message
-			err := c.checkMessage(msg.Code, view)
-			if err == nil {
-				logger.Trace("Post backlog event", "msg", msg)
-
-				go c.sendEvent(backlogEvent{
-					src: src,
-					msg: msg,
-				})
-			} else if err == errFutureMessage {
-				logger.Trace("Stop processing backlog", "msg", msg)
-				backlog.Push(msg, prio)
-				isFuture = true
-			} else {
-				logger.Trace("Skip the backlog event", "msg", msg, "err", err)
-			}
+				if err == errFutureMessage {
+					logger.Warn("Future message in backlog for seq, pushing back to the backlog", "msg", msg)
+					return true
+				}
 
+				if err == nil {
+					logger.Trace("Post backlog event", "msg", msg)
+					go c.msgProcessor(msg)
+				} else {
+					logger.Trace("Skip the backlog event", "msg", msg, "err", err)
+				}
+				return false
+			})
 		}
 	}
 }
 
 func toPriority(msgCode uint64, view *istanbul.View) int64 {
 	if msgCode == istanbul.MsgRoundChange {
-		// For istanbul.MsgRoundChange, set the message priority based on its sequence
-		return -int64(view.Sequence.Uint64() * 1000)
+		// msgRoundChange comes first
```

Comment:
This seems like a departure?

Reply:
Oh, no. It's not a departure actually. Basically the formula was:
```
priorityRoundChange = -1 * ( 1000 * sequence)
prioritiyMsg = -1 * (1000 * sequence + 10 * round + msgCodePriority)
```
but since we have one queure per sequence, we can remove that term, so we now have:
```
priorityRoundChange = 0
prioritiyMsg = -1 * (10 * round + msgCodePriority)
```
The order relation is still the same

Reply:
👍